### PR TITLE
Vehicle part installation filter[CR][WIP]

### PR DESCRIPTION
**Pull request by epsilon-phase**
Addresses #18447 

I think I should add another search qualifier than the name or the flag, but I don't know what would be useful.

To elaborate on how this works, at the moment, you have to go into the part installation screen and then set the filter.

### CODE REVIEW

File: src/veh_interact.cpp [hunk 1]
```diff
@@ -1494,6 +1515,88 @@ void veh_interact::move_cursor (int dx, int dy)
     display_mode (' ');
 }
 
+// <<<<<<< HEAD
```
Comment (mugling) [inline src/veh_interact.cpp]: Looks like bad merge?

**Comment by jaked122** [comment]
Oh no. I thought I got all of those.

Oh wait, the reason I commented those out was to test it.

I'll just rip it out, it works fine.

File: src/veh_interact.cpp [hunk 2]
```diff
@@ -694,6 +701,20 @@ void veh_interact::do_install()
         bool can_install = can_install_part();
 
         const std::string action = main_context.handle_input();
+        if (action == "FILTER" ){
+            filter = string_input_popup( _("Set Filter"),
+                                        50, filter,
+                                        _("{f:} to filter by part flag, {pgt} and {plt} to filter by power, combine with ',', otherwise it searches by name"),
```
Comment (mugling) [inline src/veh_interact.cpp]: Feeping creaturism. Search by name would be adequate here. Anything more is of marginal benefit vs the bloating of the implementation.

Comment (jaked122) [inline src/veh_interact.cpp]: Eh, I kinda like it, but if you think it's not worth the trouble, then perhaps it should wait until there's some sort of unified way to go about this.

**Comment by jaked122** [comment]
@mugling, I removed the flag based searching as you asked, and I fixed the merge.

File: src/veh_interact.h [hunk 3]
```diff
@@ -172,6 +172,10 @@ class veh_interact
         void allocate_windows();
         void do_main_loop();
         void deallocate_windows();
+
+        /*Filter stuff*/
+        bool filter_function( const vpart_info * ) const;
```
Comment (mugling) [inline src/veh_interact.h]: Why a class member?

Comment (jaked122) [inline src/veh_interact.h]: I wasn't sure where to put the function, though I don't think it's necessary if I don't have the complicated filter function that I previously had.

File: src/veh_interact.h [hunk 4]
```diff
@@ -172,6 +172,9 @@ class veh_interact
         void allocate_windows();
         void do_main_loop();
         void deallocate_windows();
+
+        /*Filter stuff*/
+        std::string filter;
```
Comment (mugling) [inline src/veh_interact.h]: Can you avoid this as well?

File: src/veh_interact.cpp [hunk 5]
```diff
@@ -34,6 +34,8 @@
 #define ISNAN std::isnan
 #endif
 
+bool lcmatch(const std::string&, const std::string&);
```
Comment (mugling) [inline src/veh_interact.cpp]: Do you need this? Perhaps you need to include a header instead?

File: src/veh_interact.h (hunk 4 above)
Comment (jaked122) [inline src/veh_interact.h]: You want me to get rid of the filter in the interface?

I'm really not sure, that's part of the state, isn't it?

File: src/veh_interact.cpp [hunk 6]
```diff
@@ -34,6 +34,7 @@
 #define ISNAN std::isnan
 #endif
 
+
```
Comment (mugling) [inline src/veh_interact.cpp]: Spurious newline

File: src/veh_interact.cpp [hunk 7]
```diff
@@ -127,7 +128,7 @@ veh_interact::veh_interact( vehicle &veh, int x, int y )
     main_context.register_action("NEXT_TAB");
     main_context.register_action("CONFIRM");
     main_context.register_action("HELP_KEYBINDINGS");
-
+    main_context.register_action("FILTER");
     countDurability();
```
Comment (mugling) [inline src/veh_interact.cpp]: Try and preserve existing formatting (line break before next logical block)

File: src/veh_interact.h (hunk 4 above)
Comment (mugling) [inline src/veh_interact.h]: Well you only need that state whilst your filter runs. Try and keep as much as possible out of the public header

File: src/veh_interact.cpp [hunk 8]
```diff
@@ -663,13 +667,16 @@ void veh_interact::do_install()
...
-        for (size_t i=1; i < tab_filters.size()-1; i++ ) {
+                                                   (part.location.empty() && part.has_flag("FUEL_TANK")); };
+
+    tab_filters[tab_filters.size()-2] = [&](const vpart_info *part) { // Other: everything that's not in the other filters
+        for (size_t i=1; i < tab_filters.size()-2; i++ ) {
             if( tab_filters[i](part) ) return false;
         }
         return true; };
-
+    tab_filters[7] = [&](const vpart_info *p){//The user specified filter
```
Comment (mugling) [inline src/veh_interact.cpp]: Would be cleaner all on one line with the comment on the line above it. Sorry for all the corrections but its important the code style is consistent for the project

File: src/veh_interact.cpp [hunk 9]
```diff
@@ -663,13 +667,16 @@ void veh_interact::do_install()
                                                    part.has_flag(VPFLAG_CONTROLS) ||
                                                    part.location == "fuel_source" ||
                                                    part.location == "on_battery_mount" ||
-                                                   part.location.empty(); };
-    tab_filters[tab_filters.size()-1] = [&](const vpart_info *part) { // Other: everything that's not in the other filters
-        for (size_t i=1; i < tab_filters.size()-1; i++ ) {
+                                                   (part.location.empty() && part.has_flag("FUEL_TANK")); };
+
+    tab_filters[tab_filters.size()-2] = [&](const vpart_info *part) { // Other: everything that's not in the other filters
```
Comment (mugling) [inline src/veh_interact.cpp]: Comment is too long to be inline, place on line above

File: src/veh_interact.cpp [hunk 10]
```diff
@@ -577,6 +578,7 @@ void veh_interact::do_install()
     display_mode('i');
     werase (w_msg);
     int msg_width = getmaxx(w_msg);
+    std::string filter;
```
Comment (mugling) [inline src/veh_interact.cpp]: Define this near where you use it. The days of C89 are long gone

File: src/veh_interact.cpp [hunk 11]
```diff
@@ -694,6 +701,20 @@ void veh_interact::do_install()
         bool can_install = can_install_part();
 
         const std::string action = main_context.handle_input();
+        if (action == "FILTER" ){
+            filter = string_input_popup( _("Set Filter"),
```
Comment (mugling) [inline src/veh_interact.cpp]: x position is off by one position too far left

File: src/veh_interact.cpp [hunk 12]
```diff
@@ -694,6 +701,20 @@ void veh_interact::do_install()
...
+                                        50, filter,
+                                        "",
+                                        _("Filter"), 100, false );
+            tab = 7;//Move to the user filter tab.
+            display_grid();
+            display_stats();
```
Comment (mugling) [inline src/veh_interact.cpp]: Blueprint display disappears after first search

**Comment by mugling** [comment]
Apart from this appears to work and good to go

File: src/veh_interact.cpp (hunk 8 above)
Comment (jaked122) [inline src/veh_interact.cpp]: No, I understand, this was a consistent issue for me last time I tried to commit too. 

I'm not terribly good at it, so I really appreciate you going through this with me.

**Comment by jaked122** [comment]
@mugling I'm glad to hear that. 

I think I've fixed the vehicle blueprint not being redrawn after the filter window is closed.

**Comment by mugling** [comment]
Your first PR? You should advertise this new feature on the forums as it's a welcome addition.

**Comment by jaked122** [comment]
@mugling I'm also responsible for the plows, reapers, and street sweeper scoops. :)

**Pull request merged**